### ISSUE TITLE
Add support for Azure App Roles

### ISSUE THREAD

**Issue by reasland** [jenkinsci/azure-ad-plugin#129]
### Feature Request

I currently have a fork of azure-ad-plugin that I'm trying to implement this when I have time (I am not a Java dev by far however)
https://github.com/reasland/azure-ad-plugin/tree/dev/src/main/java/com/microsoft/jenkins/azuread

But I want to throw out a feature request to support Azure AD App Roles in addition to the current Security Groups

https://docs.microsoft.com/en-us/azure/active-directory/develop/howto-add-app-roles-in-azure-ad-apps

Basically App Roles (maybe still in preview) are a App specific Security Group that is limited to only the App, so it does not require adding Microsoft Graph permissions to extend support for giving the App permissions to look at all Security Groups, etc...

**Comment by KalleOlaviNiemitalo** [comment]
According to [Microsoft identity platform ID tokens](https://docs.microsoft.com/en-us/azure/active-directory/develop/id-tokens#payload-claims), the value of the "roles" claim is an array of strings, so I think it could be extracted the same way as "groups" here: <https://github.com/jenkinsci/azure-ad-plugin/blob/437429002c6b313b8d3a1c6903c13726b7a7bea6/src/main/java/com/microsoft/jenkins/azuread/AzureAdUser.java#L95>

A minimal implementation without support for Microsoft Graph API could perhaps just put the role IDs in the same list as group IDs, and they could then be used for authorization within Jenkins.

However, I have not verified that app roles really go to the "roles" claim. Maybe there is some easy way to log all the claims in order to verify this, before attempting to implement anything there.

**Comment by KalleOlaviNiemitalo** [comment]
Should permission entries for Azure app roles have AuthorizationType.GROUP ([source](https://github.com/jenkinsci/matrix-auth-plugin/blob/matrix-auth-3.0/src/main/java/org/jenkinsci/plugins/matrixauth/AuthorizationType.java#L37)) or something new?

**Comment by timja** [comment]
Group I think is what it maps to here, e.g. not sure if you get a uuid or a display name but if `12313-213121-12313-1231` is returned, you're using that to allow multiple people to get say the `Job/Build` permission

**Comment by seeruebe** [comment]
Hello, does anybody know, whether that feature is implemented by now? I see, it is still "Open", but may be with another approach.
We also would like to use the app roles, because my enterprise does not allow the needed api permissions needed for the Azure AD Matrix-based security authorization.

**Comment by seeruebe** [comment]
Thank you for your prompt reply.

**Comment by lchforcom** [comment]
It would be really helpful if Azure App Roles are supported. Compare to Azure Group, there is company resitriction that Group.Read.All is not allowed to use. Our IT colleage response is the scope "Group.Read.All" is blacklisted and won't be granted as it clould includes access to M365 related services like group chats and shared files.

**Comment by timja** [comment]
Agreed it would be helpful.

In case its useful you don't have to grant the permission for it to work although there are a couple of caveats:

You can also configure the app registration to emit group claims as part of login:
https://learn.microsoft.com/en-us/entra/identity/hybrid/connect/how-to-connect-fed-group-claims#configure-the-microsoft-entra-application-registration-for-group-attributes

https://github.com/jenkinsci/azure-ad-plugin/blob/master/src/main/java/com/microsoft/jenkins/azuread/AzureAdUser.java#L95

Groups will be supported on login but there will be no autocomplete in the UI for finding them and API token authentication that relies on group permissions won't work.

**Comment by O5ten** [comment]
Just migrated from the jenkins SAML plugin (that has other defects we can no longer live with) and noticed that i can't configure the plugin to use roles instead of groups. I don't need the autocompletion tbh so i'd be fine with lacking that feature. 

I'm a Java developer, so i can perhaps help out unless someone has made some previous work?
